src/devices/bus/vcs_ctrl/trakball.cpp

```diff
@@ -0,0 +1,122 @@
+
+static INPUT_PORTS_START(atari_trakball)
+	PORT_START(TRAKBALL_BUTTON_TAG) /* Trak-ball - button */
+	PORT_BIT( 0x20, IP_ACTIVE_HIGH, IPT_BUTTON1) PORT_WRITE_LINE_MEMBER(atari_trakball_device, trigger_w)
+	PORT_BIT( 0xdf, IP_ACTIVE_HIGH, IPT_UNUSED )
+
+	PORT_START(TRAKBALL_XAXIS_TAG) /* Trak-ball - X AXIS */
+	PORT_BIT( 0xff, 0x00, IPT_TRACKBALL_X) PORT_SENSITIVITY(80) PORT_KEYDELTA(0) PORT_PLAYER(1)
+
+	PORT_START(TRAKBALL_YAXIS_TAG) /* Trak-ball - Y AXIS */
+	PORT_BIT( 0xff, 0x00, IPT_TRACKBALL_Y) PORT_SENSITIVITY(80) PORT_KEYDELTA(0) PORT_PLAYER(1)
+INPUT_PORTS_END
+
+
+//**************************************************************************
+//  LIVE DEVICE
+//**************************************************************************
+
+//-------------------------------------------------
+//  atari_cx85_device - constructor
+//-------------------------------------------------
+
+atari_trakball_device::atari_trakball_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: device_t(mconfig, ATARI_TRAKBALL, tag, owner, clock)
+	, device_vcs_control_port_interface(mconfig, *this)
+	, m_trakballb(*this, TRAKBALL_BUTTON_TAG)
+	, m_trakballxy(*this, { TRAKBALL_XAXIS_TAG, TRAKBALL_YAXIS_TAG })
+	, m_last_pos{ TRAKBALL_POS_UNINIT, TRAKBALL_POS_UNINIT }
+	, m_last_pos_sent{ 0, 0 }
+{
+}
+
+//-------------------------------------------------
+//  device_input_ports - device-specific input ports
+//-------------------------------------------------
+
+ioport_constructor atari_trakball_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME(atari_trakball);
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void atari_trakball_device::device_start()
+{
+}
+
+
+//-------------------------------------------------
+//  vcs_joy_r - read digital inputs
+//-------------------------------------------------
+
+u8 atari_trakball_device::vcs_joy_r()
+{
+	int diff_pos[2] = {0, 0};
+	int cur_pos[2] = {0, 0};
+	u8 vcs_joy_return = 0;
+
+	for (int axis = 0; axis < 2; axis++) {
+		cur_pos[axis] = m_trakballxy[axis]->read();
+		if (m_last_pos[axis] == TRAKBALL_POS_UNINIT) {
+			if (!machine().side_effects_disabled()) {
+				m_last_pos[axis] = cur_pos[axis];
+			}
+		}
+		diff_pos[axis] = cur_pos[axis] - m_last_pos[axis];
+		// wrap-around the position
+		if (diff_pos[axis] > 0x7f) {
+			diff_pos[axis] -= 0x100;
+		} else if (diff_pos[axis] < -0x80) {
+			diff_pos[axis] += 0x100;
+		}
+		if (!machine().side_effects_disabled()) {
+			m_last_pos[axis] = cur_pos[axis];
+			if (diff_pos[axis]) m_last_pos_sent[axis] = !m_last_pos_sent[axis];
+		}
+	}
+
+	vcs_joy_return =
+		m_trakballb->read() |
+		((diff_pos[0] > 0) ? 0x01 : 0x00) |
+		((m_last_pos_sent[0] > 0) ? 0x02 : 0x00) |
+		((diff_pos[1] > 0) ? 0x04 : 0x00) |
+		((m_last_pos_sent[1] > 0) ? 0x08 : 0x00);
+
+	return vcs_joy_return;
+}
```

Comment:
Given the wide variety of machines using the VCS control port, it really isn’t safe to make assumptions about when it’s going to be read.  You really need to use `INPUT_CHANGED_MEMBER` to get notifications when the axes are updated and use timers to generate the quadrature signals.

Reply:
I agree that my first submission was not realistic - mainly due to this line "if (diff_pos[axis]) m_last_pos_sent[axis] = !m_last_pos_sent[axis];" which is affected by the machine-dependent rate that vcr_joy_r() is called.

While working on the updates I just checked-in now, I had initially added INPUT_CHANGED_MEMBER and removed any trackball read() calls from vcr_joy_r(), but noticed that INPUT_CHANGED_MEMBER would only get called about once per 16ms, whereas, on the A800, vcs_joy_r gets invoked several times in 1ms.  Using just the INPUT_CHANGED_MEMBER function for the read() calls makes the trackball less controllable due to a loss of granularity.  Note that I did not add any timers for that test.  I assume the timers are meant to be used if a clock signal is being sent from the home computer to the trackball.  If so, then this is not the case for this device.

The vic20 appears to call vcr_joy_r() at a much lower rate than the a800, so I agree that those cases need to be handled as well.  What I have changed for this commit is:
- relying now on the input device's positional port-bits to determine the X/Y motion values instead of simply toggling the values each time there was movement detected
- added an INPUT_CHANGED_MEMBER so that trackball X/Y directional changes are not missed on systems that do not call vcs_joy_r() in a consistent and/or frequent manner
- kept logic to read and track X/Y movement and directional changes under vcs_joy_r() as well, so that granularity is not lost on systems that call it at much faster rate than INPUT_CHANGED_MEMBER would be called

Reply:
It’s still completely unrealistic, though.  The quadrature outputs are connected directly to the connector, and will update at a rate that depends on the speed you move the trackball.  Reading the port should have no side effects on the connected device as it isn’t strobed in any way.

You only get updates for inputs once for each frame interval for the first emulated screen (limitation of MAME).  To simulate the quadrature output, you need to use a timer to update it at a rate proportional to the accumulated delta value.

The Namco System 22 driver uses a pair of timers for the trackball quadrature signals: https://github.com/mamedev/mame/blob/mame0270/src/mame/namco/namcos22.cpp#L2906

The Mac 128K/512K/Plus updates the mouse quadrature signals while it’s updating the video output, using the video line frequency as a crude timer: https://github.com/mamedev/mame/blob/mame0270/src/mame/apple/mac128.cpp